node/src/components/binary_port.rs

```diff
@@ -0,0 +1,887 @@
+        + From<ChainspecRawBytesRequest>
+        + Send,
+{
+    let version = effect_builder.get_protocol_version().await;
+    loop {
+        let Some(incoming_request) = server.next_request().await? else {
+            debug!("remote party closed the connection");
+            return Ok(());
+        };
+
+        let Some(payload) = incoming_request.payload() else {
+            return Err(Error::NoPayload);
+        };
+
+        let resp = handle_payload(effect_builder, payload, version).await?;
+        let resp_and_payload = BinaryResponseAndRequest::new(resp, payload);
+        incoming_request.respond(Some(Bytes::from(ToBytes::to_bytes(&resp_and_payload)?)))
+    }
+}
+
+async fn handle_payload<REv>(
+    effect_builder: EffectBuilder<REv>,
+    payload: &[u8],
+    protocol_version: ProtocolVersion,
+) -> Result<BinaryResponse, Error>
+where
+    REv: From<Event>,
+{
+    let (header, remainder) = BinaryRequestHeader::from_bytes(payload)?;
+
+    if !header
+        .protocol_version()
+        .is_compatible_with(&protocol_version)
+    {
+        return Ok(BinaryResponse::new_error(
+            binary_port::ErrorCode::UnsupportedProtocolVersion,
+            protocol_version,
+        ));
+    }
+
+    // we might receive a request added in a minor version if we're behind
+    let (tag, _) = u8::from_bytes(remainder)?;
+    if BinaryRequestTag::try_from(tag).is_err() {
+        return Ok(BinaryResponse::new_error(
+            binary_port::ErrorCode::UnsupportedRequest,
+            protocol_version,
+        ));
+    }
+
+    let request = bytesrepr::deserialize_from_slice(remainder)?;
+    let resp = effect_builder
+        .make_request(
+            |responder| Event::HandleRequest { request, responder },
+            QueueKind::Regular,
+        )
+        .await;
+    Ok(resp)
+}
+
+async fn handle_client<REv>(
+    addr: SocketAddr,
+    mut client: TcpStream,
+    effect_builder: EffectBuilder<REv>,
+    config: Arc<Config>,
+    permit: OwnedSemaphorePermit,
+) where
+    REv: From<Event>
+        + From<StorageRequest>
+        + From<ContractRuntimeRequest>
+        + From<AcceptTransactionRequest>
+        + From<NetworkInfoRequest>
+        + From<ReactorInfoRequest>
+        + From<ConsensusRequest>
+        + From<BlockSynchronizerRequest>
+        + From<UpgradeWatcherRequest>
+        + From<ChainspecRawBytesRequest>
+        + Send,
+{
+    let (reader, writer) = client.split();
+    let (client, server) = new_rpc_builder(&config).build(reader, writer);
+
+    if let Err(err) = client_loop(server, effect_builder).await {
+        // Low severity is used to prevent malicious clients from causing log floods.
+        info!(%addr, %err, "binary port client handler error");
+    }
+
+    // We are a server, we won't make any requests of our own, but we need to keep the client
+    // around, since dropping the client will trigger a server shutdown.
+    drop(client);
+    drop(permit);
```

Comment:
this comment is here because the Juliet builder API gives us a pair of a client and a server, even when we don't want to make requests, so I think it makes sense
the `drop`s are indeed unnecessary, I think they've been added to silence the unused variable warning but we can just add underscores to the names if you prefer that

Reply:
Ah, right.  Yes the comment makes sense now.

I think my confusion was due to the comment saying we shouldn't drop the client, but then we immediately drop the client :)  I guess I'd have placed the comment right before the call to `new_rpc_builder`.  Also, I think it's more usual to silence the unused warning by adding the leading underscore, so I'd vote for that.

Reply:
updated